README.md
VKPorter
========

Extremely small tool to export photo albums from [vk.com](https://vk.com).


## Prerequisites

Before you can start using VKPorter you have to install some python libraries if you don't have it.

### ProgressBar

[https://pypi.python.org/pypi/progressbar](https://pypi.python.org/pypi/progressbar)

    $ pip install progressbar


### Requests

[https://github.com/kennethreitz/requests](https://github.com/kennethreitz/requests)

    $ pip install requests

or, with [easy_install](http://pypi.python.org/pypi/setuptools):

    $ easy_install requests

### VK_API
[https://github.com/python273/vk_api](https://github.com/python273/vk_api)

    $ pip install vk_api

## Usage

Synopsis:

    $ vkporter.py [-h] [-v] [-o OUTPUT] username

See also `vkporter --help`.

### Examples

    $ vkporter.py username password
    
photo albums will be exported to `./exported`.


    $ vkporter.py -o ~/Documents/Exported username
    
photo albums will be exported to `~/Documents/Exported`.


CHANGELOG.md
2013.03.10, Version 0.1.2

 * Removes ProgressBar and change exporting indication
 * Updates pass asking with getpass()


porter.py
# -*- coding: utf-8 -*-
__author__ = 'amka'
__date__ = '09.03.13'

from progressbar import FileTransferSpeed, Percentage, Bar, ProgressBar
import requests
from vk_api import vk_api


def connect(login, password):
    """Initialize connection with `vk.com <https://vk.com>`_ and try to authorize user with given credentials.

    :param login: user login e. g. email, phone number
    :type login: str
    :param password: user password
    :type password: str

    :return: :class:`vk_api.VkApi` connection
    :rtype: :class:`vk_api.VkApi`
    """
    return vk_api.VkApi(login, password)


def get_albums(connection):
    """Get albums list for currently authorized user.

    :param connection: :class:`vk_api.VkApi` connection
    :type connection: :class:`vk_api.VkApi`

    :return: list of photo albums
    :rtype: list
    """
    return []


def download(photo):
    """Download photo

    :param photo:
    """
    url = photo.get('src_xxxbig') or photo.get('src_xxbig') or photo.get('src_xbig') or photo.get('src_big')

    r = requests.get(url)
    size = int(r.headers['content-length'].strip())
    bytes = 0
    title = photo['text'] or '%s' % photo['pid']
    widgets = [title, ": ", Bar(marker="|", left="[", right=" "),
               Percentage(), " ", FileTransferSpeed(), "] ",
               ' ',
               " of {0}".format(sizeof_fmt(size))]
    pbar = ProgressBar(widgets=widgets, maxval=size).start()
    file = []
    for buf in r.iter_content(1024):
        if buf:
            file.append(buf)
            bytes += len(buf)
            pbar.update(bytes)
    pbar.finish()


def sizeof_fmt(num):
    for x in ['bytes', 'KB', 'MB', 'GB', 'TB']:
        if num < 1024.0:
            return "%3.1f %s" % (num, x)
        num /= 1024.0

if __name__ == '__main__':
    login = raw_input(u"Enter your username: ")
    password = raw_input(u"Enter your password: ")
    connection = connect(login, password)

    # Connect to vk.com
    # Get list of user albums
    # For every album do the follow:
    # # create folder if not exists
    # # get list of photos
    # # download photo and put it in album folder
    # # do it while have photos
    # do it while have albums

vkporter.py
#!/usr/bin/env python

"""
    :mod:`vkporter`
    ~~~~~~~~~~~~~~~

    A micro tool for export photo albums from `vk.com <https://vk.com>`_.
     It's based on `VK_API <https://github.com/python273/vk_api>`_
     by Kirill Python <mikeking568@gmail.com>,
     `Requests <python-requests.org>`_
     and `ProgressBar <https://code.google.com/p/python-progressbar/>`_.

    :copyright: (c) 2013 by Andrey Maksimov.
    :license: BSD, see LICENSE for more details.
"""

__author__ = 'Andrey Maksimov <meamka@me.com>'
__date__ = '09.03.13'
__version__ = '0.1.1'

import argparse
import datetime
from getpass import getpass
import os
import time
import sys

try:
    import requests
except ImportError:
    print("Cannot find 'requests' module. Please install it and try again.")
    sys.exit(0)

try:
    from vk_api import VkApi
except ImportError:
    print("Cannot find 'vk_api' module. Please install it and try again.")
    sys.exit(0)


def connect(login, password):
    """Initialize connection with `vk.com <https://vk.com>`_ and try to authorize user with given credentials.

    :param login: user login e. g. email, phone number
    :type login: str
    :param password: user password
    :type password: str

    :return: :mod:`vk_api.vk_api.VkApi` connection
    :rtype: :mod:`VkApi`
    """
    return VkApi(login, password)


def get_albums(connection):
    """Get albums list for currently authorized user.

    :param connection: :class:`vk_api.vk_api.VkApi` connection
    :type connection: :class:`vk_api.vk_api.VkApi`

    :return: list of photo albums or ``None``
    :rtype: list
    """
    try:
        return connection.method('photos.getAlbums')
    except Exception as e:
        print(e)
        return None


def get_photos(connection, album_id):
    """Get photos list for selected album.

    :param connection: :class:`vk_api.vk_api.VkApi` connection
    :type connection: :class:`vk_api.vk_api.VkApi`
    :param album_id: album identifier returned by :func:`get_albums`
    :type album_id: int

    :return: list of photo albums or ``None``
    :rtype: list
    """
    try:
        return connection.method('photos.get', {'aid': album_id})
    except Exception as e:
        print(e)
        return None


def download(photo, output):
    """Download photo

    :param photo:
    """
    url = photo.get('src_xxxbig') or photo.get('src_xxbig') or photo.get('src_xbig') or photo.get('src_big')

    r = requests.get(url)
    title = photo['pid']
    with open(os.path.join(output, '%s.jpg' % title), 'wb') as f:
        for buf in r.iter_content(1024):
            if buf:
                f.write(buf)


def sizeof_fmt(num):
    """Small function to format numbered size to human readable string

    :param num: bytes to format
    :type num: int

    :return: human readable size
    """
    for x in ['bytes', 'KB', 'MB', 'GB', 'TB']:
        if num < 1024.0:
            return "%3.1f %s" % (num, x)
        num /= 1024.0


if __name__ == '__main__':

    # Connect to vk.com
    # Get list of user albums
    # For every album do the follow:
    # # create folder if not exists
    # # get list of photos
    # # download photo and put it in album folder
    # # do it while have photos
    # do it while have albums

    parser = argparse.ArgumentParser(description='', version='%(prog)s ' + __version__)
    parser.add_argument('username', help='vk.com username')
    # parser.add_argument('password', help='vk.com username password')
    parser.add_argument('-o', '--output', help='output path to store photos',
                        default=os.path.abspath(os.path.join(os.path.dirname(__file__), 'exported')))

    args = parser.parse_args()

    # expand user path if necessary
    if args.output.startswith('~'):
        args.output = os.path.expanduser(args.output)

    start_time = datetime.datetime.now()
    try:
        password = getpass("Password: ")
        if not password:
            print('Password too short')
            sys.exit(0)

        # Initialize vk.com connection
        connection = connect(args.username, password)

        # Request list of photo albums
        albums = get_albums(connection)
        print("Found %s album%s:" % (len(albums), 's' if len(albums) > 1 else ''))
        ix = 0
        for album in albums:
            print('%3d. %-40s %4s item%s' % (
            ix + 1, album['title'], album['size'], 's' if int(album['size']) > 1 else ''))
            ix += 1

        # Sleep to prevent max request count
        time.sleep(1)

        if not os.path.exists(args.output):
            os.makedirs(args.output)

        prev_s_len = 0  # A length of the previous output line.
        for album in albums:
            response = get_photos(connection, album['aid'])
            output = os.path.join(args.output, album['title'])
            if not os.path.exists(output):
                os.makedirs(output)

            processed = 0

            for photo in response:
                percent = round(float(processed) / float(len(response)) * 100, 2)
                output_s = "\rExporting %s... %s of %s (%2d%%)" % (album['title'], processed, len(response), percent)
                # Pad with spaces to clear the previous line's tail.
                # It's ok to multiply by negative here.
                output_s += ' '*(prev_s_len - len(output_s))
                sys.stdout.write(output_s)
                prev_s_len = len(output_s)
                sys.stdout.flush()

                download(photo, output)
                processed += 1

    except Exception as e:
        print(e)
        sys.exit(1)

    except KeyboardInterrupt:
        print('VKPorter exporting stopped by keyboard')
        sys.exit(0)

    finally:
        print("Done in %s" % (datetime.datetime.now() - start_time))
